Add button entities to bmw_connected_drive

rikroe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
`bmw_connected_drive.*` services are deprecated. Please use the new button entities with the `button.press` service.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds existing services as buttons.

~As all services were just one-shot clicks without any data, the services are removed. This enabled cleaning up `__init__.py`.~ (Services are still included and will be removed in a later release).


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20959

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rikroe committed:
Add Button platform to BMW Connected Drive

rikroe committed:
Remove services

rikroe committed:
Add .coveragerc exception

epenet commented:
There is usually a deprecation period implemented with a warning (see Renault). I suggest you only remove the services in 2022.04 (after two releases)

filcole commented:
Just noticed that BMW has two buttons to turn on/off ac. Nissan leaf has a switch for the AC.

rikroe committed:
Revert "Remove services"

This reverts commit b58d0fe7637fc3b59cfd2d54965c91b633dd1075.

rikroe committed:
Add deprecation warning for BMW services

rikroe commented:
@epenet: thanks, you're right. Reverted the change and added a warning.

@filcole: Yes, that is correct. Currently we're unable to identify if a car has the ability to stop AC or if the AC is actually running (getting the state). Therefore it's just two fire-and-forget buttons. 
If we figure out one or the other (or both), then I could think of creating conditional buttons - if stopping is possible, then show stop, if both possible + state is available, then no buttons but switch, ...

gerard33 commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+"""Support for Renault button entities."""
```
BMW

epenet replied:
```suggestion
"""Support for BMW connected drive button entities."""
```

rikroe committed:
Update bmw_connected_drive/button.py

Co-authored-by: Gerard <gerard33@users.noreply.github.com>

lawtancool commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+    BMWButtonEntityDescription(
+        key="refresh",
+        icon="mdi:refresh",
+        name="Refresh from API",
```
Renaming to "Refresh data from BMW API" might make the function of this button slightly more clear.

rikroe replied:
Thats true. Would you prefer "Refresh from BMW API" or "Refresh from BMW servers"?

filcole replied:
For Nissan Leaf and Renault the docs mention 'cloud service', because there's not an official API.

rikroe replied:
`cloud` does make sense in this case!

lawtancool commented on homeassistant/components/bmw_connected_drive/__init__.py:
```diff
@@ -223,6 +224,10 @@ def setup_account(
     def execute_service(call: ServiceCall) -> None:
         """Execute a service for a vehicle."""
+        _LOGGER.warning(
+            "BMW Connected Drive services are deprecated. Please migrate to the dedicated button entities"
```
Adding a link to the updated integration documentation on the Home Assistant website in the warning message would help less advanced users understand what is wrong.

```"BMW Connected Drive services are deprecated and will be removed in a future release. Please migrate to the dedicated button entities. See https://www.home-assistant.io/integrations/bmw_connected_drive/#buttons"```

rikroe committed:
Add link to documentation to warning message

rikroe committed:
Rename API to cloud

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+    """Class describing BMW button entities."""
+
+    enabled_when_read_only: bool = False
+    remote_function: str | None = None
+    account_function: str | None = None
```
```suggestion
    remote_function: Callable[[ConnectedDriveVehicle], Callable[[], None]] | None = None
    account_function: Callable[[BMWConnectedDriveAccount], Callable[[], None]] | None = None
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        key="light_flash",
+        icon="mdi:car-light-alert",
+        name="Flash Lights",
+        remote_function="trigger_remote_light_flash",
```
```suggestion
        remote_function=lambda vehicle: vehicle.remote_services.trigger_remote_light_flash,
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        key="sound_horn",
+        icon="mdi:bullhorn",
+        name="Sound Horn",
+        remote_function="trigger_remote_horn",
```
```suggestion
        remote_function=lambda lambda vehicle: vehicle.remote_services.trigger_remote_horn,
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        key="activate_air_conditioning",
+        icon="mdi:hvac",
+        name="Activate Air Conditioning",
+        remote_function="trigger_remote_air_conditioning",
```
```suggestion
        remote_function=lambda vehicle: vehicle.remote_services.trigger_remote_air_conditioning,
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        key="deactivate_air_conditioning",
+        icon="mdi:hvac-off",
+        name="Deactivate Air Conditioning",
+        remote_function="trigger_remote_air_conditioning_stop",
```
```suggestion
        remote_function=lambda vehicle: vehicle.remote_services.trigger_remote_air_conditioning_stop,
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        key="find_vehicle",
+        icon="mdi:crosshairs-question",
+        name="Find Vehicle",
+        remote_function="trigger_remote_vehicle_finder",
```
```suggestion
        remote_function=lambda vehicle: vehicle.remote_services.trigger_remote_vehicle_finder,
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        key="refresh",
+        icon="mdi:refresh",
+        name="Refresh from cloud",
+        account_function="update",
```
```suggestion
        account_function=lambda account: account.update,
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+        self._attr_name = f"{vehicle.name} {description.name}"
+        self._attr_unique_id = f"{vehicle.vin}-{description.key}"
+
+    async def async_press(self) -> None:
```
```suggestion
   def press(self) -> None:
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+
+    async def async_press(self) -> None:
+        """Process the button press."""
+        if self.entity_description.remote_function:
+            function_call = getattr(
+                self._vehicle.remote_services,
+                str(self.entity_description.remote_function),
+            )
```
```suggestion
        if function_call := self.entity_description.remote_function(self._vehicle):
           function_call()
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+                self._vehicle.remote_services,
+                str(self.entity_description.remote_function),
+            )
+        elif self.entity_description.account_function:
+            function_call = getattr(
+                self._account, str(self.entity_description.account_function)
+            )
```
```suggestion
        elif function_call := self.entity_description.account_function(self._account):
            function_call()
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,125 @@
+            function_call = getattr(
+                self._account, str(self.entity_description.account_function)
+            )
+
+        await self.hass.async_add_executor_job(function_call)
```
```suggestion
```

frenck reviewed (changes requested):
A suggestion for an alternative approach, that removes the need for doing getatrrs and such, and works correctly with typing as well.

rikroe committed:
Use lambda functions and sync press

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

rikroe committed:
Fix using lambda functions

rikroe commented:
Thanks frenck, lambda functions do make more sense. 

I only had to change the return value of the callables to `Any` to make mypy happy. Not sure if there is a better way, but with the return value being `Callable[[], None]` I was getting the following mypy error:
```
homeassistant/components/bmw_connected_drive/button.py:36: error: Returning Any from function declared to return "Callable[..., Any]"  [no-any-return]`
```

Not sure if this is fine for you or there is a better option.

rikroe committed:
Fix pylint

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,120 @@
+from dataclasses import dataclass
+from typing import Any
+
+from bimmer_connected.vehicle import ConnectedDriveVehicle
```
```suggestion
from bimmer_connected.remote_services import RemoteServiceStatus
from bimmer_connected.vehicle import ConnectedDriveVehicle
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,120 @@
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
```
```suggestion
```

frenck commented on homeassistant/components/bmw_connected_drive/button.py:
```diff
@@ -0,0 +1,120 @@
+    """Class describing BMW button entities."""
+
+    enabled_when_read_only: bool = False
+    remote_function: Callable[[ConnectedDriveVehicle], Any] | None = None
```
```suggestion
    remote_function: Callable[
        [ConnectedDriveVehicle], RemoteServiceStatus
    ] | None = None
```

frenck commented:
> BTW, is there a reason that pylint is not included in pre-commit? I was always checking against that but again forget pylint... :D

Yeah, it used to be too heavy to run, making committing really annoying.

rikroe committed:
Use RemoteServiceStatus instead of Any

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

rikroe commented:
Thanks for the hint, I am really getting started with strictly typed Python.

And regarding pylint: That's a very good point, it's very annoying 👍

frenck commented:
Thanks, @rikroe 👍 

Happy New Year 🍾

rikroe commented:
Thanks, same to you!!
